lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py

```diff
@@ -227,53 +253,66 @@
 from ansible.module_utils.azure_rm_common import AzureRMModuleBase, azure_id_to_dict
 
 
+def get_config_name(name, index):
+    return name or 'ipconfig{0}'.format(index)
+
+def subnet_to_dict(subnet):
+    dic = azure_id_to_dict(subnet.id)
+    return dict(
+        id=subnet.id,
+        virtual_network_name=dic.get('virtual_network_name'),
+        resource_group=dic.get('resourceGroups'),
+        name=dic.get('subnets')
+    )
+
 def nic_to_dict(nic):
-    result = dict(
+    ip_configurations = [
+        dict(
+            name=config.name,
+            private_ip_address=config.private_ip_address,
+            private_ip_allocation_method=config.private_ip_allocation_method,
+            subnet=subnet_to_dict(config.subnet),
+            primary=config.primary,
+            public_ip_address=dict(
+                id=config.public_ip_address.id,
+                name=azure_id_to_dict(config.public_ip_address.id).get('publicIPAddresses'),
+                public_ip_allocation_method=config.public_ip_address.public_ip_allocation_method
+            ) if config.public_ip_address else None
+        ) for config in nic.ip_configurations
+    ]
+    return dict(
         id=nic.id,
         name=nic.name,
         type=nic.type,
         location=nic.location,
         tags=nic.tags,
-        network_security_group=dict(),
-        ip_configuration=dict(
-            name=nic.ip_configurations[0].name,
-            private_ip_address=nic.ip_configurations[0].private_ip_address,
-            private_ip_allocation_method=nic.ip_configurations[0].private_ip_allocation_method,
-            subnet=dict(),
-            public_ip_address=dict(),
-        ),
+        network_security_group=dict(
+            id=nic.network_security_group.id,
+            name=azure_id_to_dict(nic.network_security_group.id).get('networkSecurityGroups')
+        ) if nic.network_security_group else None,
         dns_settings=dict(
             dns_servers=nic.dns_settings.dns_servers,
             applied_dns_servers=nic.dns_settings.applied_dns_servers,
             internal_dns_name_label=nic.dns_settings.internal_dns_name_label,
             internal_fqdn=nic.dns_settings.internal_fqdn
         ),
+        ip_configurations=ip_configurations,
+        ip_configuration=ip_configurations[0] if len(ip_configurations) == 1 else None,  # for compatiable issue, keep this field
```

Comment:
Can you not just set a list of 1 dict in `ip_configurations`?

Reply:
Actually I don't like the `ip_configuration` at all.

It is a field already exist in the response. So consider the compatible issue, I keep this field here. If there is a deprecation, it should be removed immediately.